File: src/lua/wave/wave-manager.lua
--[[
--------------------------------------------------
波次管理模块 - Wave Manager Module
Fighter Game Lua Script Module

功能概述:
- 波次状态管理（准备、进行中、暂停、完成）
- 敌人生成配置和数量计算
- Boss和精英波次触发逻辑
- 难度曲线和敌人属性递增
- 波次进度追踪

作者: Fighter Game Team
版本: 1.0.0
最后更新: 2026-07-16

注意事项:
- 使用局部模块模式，避免全局变量污染
- 所有公共函数进行参数验证
- 返回值统一使用 {success, result, error} 结构
- 复用 utils.lua 和 game-config.lua 的工具函数
--------------------------------------------------
]]--

local WaveManager = {}

WaveManager.VERSION = "1.0.0"
WaveManager.AUTHOR = "Fighter Game Team"

local WaveState = {
  WAITING = "waiting",
  ACTIVE = "active",
  PAUSED = "paused",
  COMPLETED = "completed",
  FAILED = "failed"
}

local currentState = {
  waveNumber = 1,
  maxWaves = 10,
  state = WaveState.WAITING,
  enemiesSpawned = 0,
  enemiesDefeated = 0,
  enemiesRemaining = 0,
  waveStartTime = 0,
  elapsedTime = 0,
  difficulty = "normal",
  isBossWave = false,
  isEliteWave = false,
  spawnedEnemyTypes = {}
}

local difficultyMultipliers = {
  easy = { health = 0.8, damage = 0.8, speed = 0.8, count = 0.8 },
  normal = { health = 1.0, damage = 1.0, speed = 1.0, count = 1.0 },
  hard = { health = 1.2, damage = 1.2, speed = 1.1, count = 1.2 },
  nightmare = { health = 1.5, damage = 1.5, speed = 1.3, count = 1.5 }
}

local enemyTemplates = {
  basic = { type = "basic", health = 50, damage = 10, speed = 2.0, score = 100 },
  fast = { type = "fast", health = 30, damage = 8, speed = 4.0, score = 150 },
  tank = { type = "tank", health = 100, damage = 15, speed = 1.0, score = 250 },
  shooter = { type = "shooter", health = 40, damage = 12, speed = 1.5, score = 200 },
  elite = { type = "elite", health = 200, damage = 25, speed = 2.5, score = 500 },
  boss = { type = "boss", health = 1000, damage = 30, speed = 1.5, score = 2000 }
}

function WaveManager.getWaveState()
  return {
    success = true,
    state = {
      waveNumber = currentState.waveNumber,
      maxWaves = currentState.maxWaves,
      currentState = currentState.state,
      enemiesSpawned = currentState.enemiesSpawned,
      enemiesDefeated = currentState.enemiesDefeated,
      enemiesRemaining = currentState.enemiesRemaining,
      elapsedTime = currentState.elapsedTime,
      difficulty = currentState.difficulty,
      isBossWave = currentState.isBossWave,
      isEliteWave = currentState.isEliteWave,
      progress = WaveManager.calculateProgress()
    }
  }
end

function WaveManager.setDifficulty(difficulty)
  if type(difficulty) ~= "string" then
    return { success = false, error = "difficulty must be a string" }
  end

  local validDifficulties = { easy = true, normal = true, hard = true, nightmare = true }
  if not validDifficulties[difficulty] then
    return { success = false, error = "invalid difficulty: " .. difficulty }
  end

  currentState.difficulty = difficulty
  return { success = true }
end

function WaveManager.setMaxWaves(maxWaves)
  if type(maxWaves) ~= "number" or maxWaves < 1 then
    return { success = false, error = "maxWaves must be a positive integer" }
  end

  currentState.maxWaves = math.floor(maxWaves)
  return { success = true }
end

function WaveManager.calculateEnemyCount(waveNumber)
  if type(waveNumber) ~= "number" or waveNumber < 1 then
    return { success = false, error = "waveNumber must be a positive integer" }
  end

  local baseCount = 5
  local growthFactor = 1.1
  local difficultyMult = difficultyMultipliers[currentState.difficulty].count

  local count = math.floor(baseCount * math.pow(growthFactor, waveNumber - 1) * difficultyMult)
  
  if WaveManager.isBossWave(waveNumber) then
    count = math.max(1, math.floor(count * 0.3))
  elseif WaveManager.isEliteWave(waveNumber) then
    count = math.floor(count * 0.8)
  end

  return { success = true, count = count }
end

function WaveManager.isBossWave(waveNumber)
  if type(waveNumber) ~= "number" or waveNumber < 1 then
    return { success = false, error = "waveNumber must be a positive integer" }
  end

  return { success = true, isBoss = waveNumber % 5 == 0 }
end

function WaveManager.isEliteWave(waveNumber)
  if type(waveNumber) ~= "number" or waveNumber < 1 then
    return { success = false, error = "waveNumber must be a positive integer" }
  end

  if WaveManager.isBossWave(waveNumber).isBoss then
    return { success = true, isElite = false }
  end

  return { success = true, isElite = waveNumber % 3 == 0 }
end

function WaveManager.generateEnemyTypes(waveNumber)
  if type(waveNumber) ~= "number" or waveNumber < 1 then
    return { success = false, error = "waveNumber must be a positive integer" }
  end

  local enemyTypes = {}
  local isBoss = WaveManager.isBossWave(waveNumber).isBoss
  local isElite = WaveManager.isEliteWave(waveNumber).isElite

  if isBoss then
    enemyTypes[#enemyTypes + 1] = "boss"
  else
    local availableTypes = { "basic" }
    
    if waveNumber >= 2 then
      availableTypes[#availableTypes + 1] = "fast"
    end
    if waveNumber >= 3 then
      availableTypes[#availableTypes + 1] = "shooter"
    end
    if waveNumber >= 4 then
      availableTypes[#availableTypes + 1] = "tank"
    end
    if isElite then
      availableTypes[#availableTypes + 1] = "elite"
    end

    local countResult = WaveManager.calculateEnemyCount(waveNumber)
    if not countResult.success then
      return countResult
    end

    local count = countResult.count
    for i = 1, count do
      local randomIndex = math.random(1, #availableTypes)
      enemyTypes[#enemyTypes + 1] = availableTypes[randomIndex]
    end
  end

  return { success = true, enemyTypes = enemyTypes }
end

function WaveManager.getEnemyConfig(enemyType)
  if type(enemyType) ~= "string" then
    return { success = false, error = "enemyType must be a string" }
  end

  local template = enemyTemplates[enemyType]
  if not template then
    return { success = false, error = "unknown enemy type: " .. enemyType }
  end

  local waveMult = math.pow(1.15, currentState.waveNumber - 1)
  local diffMult = difficultyMultipliers[currentState.difficulty]

  return {
    success = true,
    config = {
      type = template.type,
      health = math.floor(template.health * waveMult * diffMult.health),
      damage = math.floor(template.damage * waveMult * diffMult.damage),
      speed = template.speed * diffMult.speed,
      score = math.floor(template.score * waveMult),
      isBoss = enemyType == "boss",
      isElite = enemyType == "elite"
    }
  }
end

function WaveManager.startWave(waveNumber)
  if type(waveNumber) ~= "number" or waveNumber < 1 then
    return { success = false, error = "waveNumber must be a positive integer" }
  end

  if waveNumber > currentState.maxWaves then
    return { success = false, error = "waveNumber exceeds maxWaves" }
  end

  local enemyTypesResult = WaveManager.generateEnemyTypes(waveNumber)
  if not enemyTypesResult.success then
    return enemyTypesResult
  end

  currentState.waveNumber = waveNumber
  currentState.state = WaveState.ACTIVE
  currentState.enemiesSpawned = 0
  currentState.enemiesDefeated = 0
  currentState.enemiesRemaining = #enemyTypesResult.enemyTypes
  currentState.waveStartTime = 0
  currentState.elapsedTime = 0
  currentState.isBossWave = WaveManager.isBossWave(waveNumber).isBoss
  currentState.isEliteWave = WaveManager.isEliteWave(waveNumber).isElite
  currentState.spawnedEnemyTypes = enemyTypesResult.enemyTypes

  return {
    success = true,
    waveNumber = waveNumber,
    enemyCount = currentState.enemiesRemaining,
    isBossWave = currentState.isBossWave,
    isEliteWave = currentState.isEliteWave,
    enemyTypes = currentState.spawnedEnemyTypes
  }
end

function WaveManager.spawnNextEnemy()
  if currentState.state ~= WaveState.ACTIVE then
    return { success = false, error = "wave is not active" }
  end

  if currentState.enemiesSpawned >= #currentState.spawnedEnemyTypes then
    return { success = false, error = "no more enemies to spawn" }
  end

  currentState.enemiesSpawned = currentState.enemiesSpawned + 1
  local enemyType = currentState.spawnedEnemyTypes[currentState.enemiesSpawned]
  
  local configResult = WaveManager.getEnemyConfig(enemyType)
  if not configResult.success then
    return configResult
  end

  return {
    success = true,
    enemy = configResult.config,
    spawnIndex = currentState.enemiesSpawned,
    totalToSpawn = #currentState.spawnedEnemyTypes
  }
end

function WaveManager.onEnemyDefeated(enemyType)
  if currentState.state ~= WaveState.ACTIVE then
    return { success = false, error = "wave is not active" }
  end

  currentState.enemiesDefeated = currentState.enemiesDefeated + 1
  currentState.enemiesRemaining = currentState.enemiesRemaining - 1

  if currentState.enemiesRemaining <= 0 then
    WaveManager.completeWave()
  end

  return {
    success = true,
    enemiesDefeated = currentState.enemiesDefeated,
    enemiesRemaining = currentState.enemiesRemaining,
    isWaveComplete = currentState.enemiesRemaining <= 0,
    score = enemyTemplates[enemyType] and enemyTemplates[enemyType].score or 100
  }
end

function WaveManager.completeWave()
  currentState.state = WaveState.COMPLETED

  return {
    success = true,
    waveNumber = currentState.waveNumber,
    enemiesDefeated = currentState.enemiesDefeated,
    elapsedTime = currentState.elapsedTime,
    isLastWave = currentState.waveNumber >= currentState.maxWaves
  }
end

function WaveManager.failWave()
  currentState.state = WaveState.FAILED

  return {
    success = true,
    waveNumber = currentState.waveNumber,
    enemiesRemaining = currentState.enemiesRemaining
  }
end

function WaveManager.pauseWave()
  if currentState.state ~= WaveState.ACTIVE then
    return { success = false, error = "wave is not active" }
  end

  currentState.state = WaveState.PAUSED
  return { success = true }
end

function WaveManager.resumeWave()
  if currentState.state ~= WaveState.PAUSED then
    return { success = false, error = "wave is not paused" }
  end

  currentState.state = WaveState.ACTIVE
  return { success = true }
end

function WaveManager.update(deltaTime)
  if type(deltaTime) ~= "number" or deltaTime < 0 then
    return { success = false, error = "deltaTime must be a non-negative number" }
  end

  if currentState.state == WaveState.ACTIVE then
    currentState.elapsedTime = currentState.elapsedTime + deltaTime
  end

  return {
    success = true,
    elapsedTime = currentState.elapsedTime,
    state = currentState.state,
    enemiesRemaining = currentState.enemiesRemaining
  }
end

function WaveManager.calculateProgress()
  if currentState.enemiesDefeated + currentState.enemiesRemaining == 0 then
    return 0
  end

  return (currentState.enemiesDefeated) / (currentState.enemiesDefeated + currentState.enemiesRemaining)
end

function WaveManager.reset()
  currentState = {
    waveNumber = 1,
    maxWaves = 10,
    state = WaveState.WAITING,
    enemiesSpawned = 0,
    enemiesDefeated = 0,
    enemiesRemaining = 0,
    waveStartTime = 0,
    elapsedTime = 0,
    difficulty = "normal",
    isBossWave = false,
    isEliteWave = false,
    spawnedEnemyTypes = {}
  }

  return { success = true }
end

function WaveManager.getNextWaveNumber()
  local nextWave = currentState.waveNumber + 1
  if nextWave > currentState.maxWaves then
    return { success = false, error = "no more waves" }
  end

  return { success = true, waveNumber = nextWave }
end

function WaveManager.getWaveScoreMultiplier()
  local baseMult = 1.0
  local waveBonus = currentState.waveNumber * 0.1
  local difficultyBonus = { easy = 0.5, normal = 1.0, hard = 1.5, nightmare = 2.0 }

  return {
    success = true,
    multiplier = baseMult + waveBonus + difficultyBonus[currentState.difficulty]
  }
end

package.preload["wave_manager_module"] = function()
  return WaveManager
end

return WaveManager

File: src/lua/wave/WaveManager.ts
import { luaEngine } from '../LuaEngine';
import type { DifficultyLevel } from '../types';

const waveScriptModules = import.meta.glob('./wave-manager.lua', { as: 'raw', eager: true });
const waveManagerScript = waveScriptModules['./wave-manager.lua'] || '';

export interface EnemyConfig {
  type: string;
  health: number;
  damage: number;
  speed: number;
  score: number;
  isBoss: boolean;
  isElite: boolean;
}

export interface WaveState {
  waveNumber: number;
  maxWaves: number;
  currentState: string;
  enemiesSpawned: number;
  enemiesDefeated: number;
  enemiesRemaining: number;
  elapsedTime: number;
  difficulty: string;
  isBossWave: boolean;
  isEliteWave: boolean;
  progress: number;
}

export class WaveManager {
  private initialized = false;

  async initialize(): Promise<void> {
    if (this.initialized) return;

    await luaEngine.initialize();

    const waveScript = await this.loadWaveScript();
    luaEngine.registerModule({ name: 'wave_manager', script: waveScript });

    this.initialized = true;
    console.log('[WaveManager] Initialized');
  }

  private async loadWaveScript(): Promise<string> {
    if (waveManagerScript) {
      return `
${waveManagerScript}

local WaveManager = require("wave_manager_module")

function getWaveState()
  return WaveManager.getWaveState().state
end

function setDifficulty(difficulty)
  local result = WaveManager.setDifficulty(difficulty)
  return result.success
end

function setMaxWaves(maxWaves)
  local result = WaveManager.setMaxWaves(maxWaves)
  return result.success
end

function calculateEnemyCount(waveNumber)
  local result = WaveManager.calculateEnemyCount(waveNumber)
  if result.success then
    return result.count
  end
  return 5
end

function isBossWave(waveNumber)
  local result = WaveManager.isBossWave(waveNumber)
  if result.success then
    return result.isBoss
  end
  return false
end

function isEliteWave(waveNumber)
  local result = WaveManager.isEliteWave(waveNumber)
  if result.success then
    return result.isElite
  end
  return false
end

function generateEnemyTypes(waveNumber)
  local result = WaveManager.generateEnemyTypes(waveNumber)
  if result.success then
    return result.enemyTypes
  end
  return {}
end

function getEnemyConfig(enemyType)
  local result = WaveManager.getEnemyConfig(enemyType)
  if result.success then
    return result.config
  end
  return nil
end

function startWave(waveNumber)
  local result = WaveManager.startWave(waveNumber)
  if result.success then
    return {
      success = true,
      waveNumber = result.waveNumber,
      enemyCount = result.enemyCount,
      isBossWave = result.isBossWave,
      isEliteWave = result.isEliteWave,
      enemyTypes = result.enemyTypes
    }
  end
  return { success = false, error = result.error }
end

function spawnNextEnemy()
  local result = WaveManager.spawnNextEnemy()
  if result.success then
    return {
      success = true,
      enemy = result.enemy,
      spawnIndex = result.spawnIndex,
      totalToSpawn = result.totalToSpawn
    }
  end
  return { success = false, error = result.error }
end

function onEnemyDefeated(enemyType)
  local result = WaveManager.onEnemyDefeated(enemyType)
  if result.success then
    return {
      success = true,
      enemiesDefeated = result.enemiesDefeated,
      enemiesRemaining = result.enemiesRemaining,
      isWaveComplete = result.isWaveComplete,
      score = result.score
    }
  end
  return { success = false, error = result.error }
end

function completeWave()
  local result = WaveManager.completeWave()
  if result.success then
    return {
      success = true,
      waveNumber = result.waveNumber,
      enemiesDefeated = result.enemiesDefeated,
      elapsedTime = result.elapsedTime,
      isLastWave = result.isLastWave
    }
  end
  return { success = false }
end

function failWave()
  local result = WaveManager.failWave()
  if result.success then
    return {
      success = true,
      waveNumber = result.waveNumber,
      enemiesRemaining = result.enemiesRemaining
    }
  end
  return { success = false }
end

function pauseWave()
  local result = WaveManager.pauseWave()
  return result.success
end

function resumeWave()
  local result = WaveManager.resumeWave()
  return result.success
end

function updateWave(deltaTime)
  local result = WaveManager.update(deltaTime)
  if result.success then
    return {
      elapsedTime = result.elapsedTime,
      state = result.state,
      enemiesRemaining = result.enemiesRemaining
    }
  end
  return { elapsedTime = 0, state = 'waiting', enemiesRemaining = 0 }
end

function resetWaveManager()
  local result = WaveManager.reset()
  return result.success
end

function getNextWaveNumber()
  local result = WaveManager.getNextWaveNumber()
  if result.success then
    return result.waveNumber
  end
  return nil
end

function getWaveScoreMultiplier()
  local result = WaveManager.getWaveScoreMultiplier()
  if result.success then
    return result.multiplier
  end
  return 1.0
end
      `;
    }

    console.warn('[WaveManager] Failed to load lua file, using fallback');
    return `
WaveManager = {
  waveNumber = 1,
  maxWaves = 10,
  state = "waiting",
  enemiesSpawned = 0,
  enemiesDefeated = 0,
  enemiesRemaining = 0,
  elapsedTime = 0,
  difficulty = "normal",
  isBossWave = false,
  isEliteWave = false
}

function getWaveState()
  return WaveManager
end

function setDifficulty(difficulty)
  WaveManager.difficulty = difficulty
  return true
end

function setMaxWaves(maxWaves)
  WaveManager.maxWaves = maxWaves
  return true
end

function calculateEnemyCount(waveNumber)
  local baseCount = 5
  local growthFactor = 1.1
  return math.floor(baseCount * math.pow(growthFactor, waveNumber - 1))
end

function isBossWave(waveNumber)
  return waveNumber % 5 == 0
end

function isEliteWave(waveNumber)
  return waveNumber % 3 == 0 and not isBossWave(waveNumber)
end

function generateEnemyTypes(waveNumber)
  local types = {}
  local count = calculateEnemyCount(waveNumber)
  for i = 1, count do
    types[i] = "basic"
  end
  return types
end

function getEnemyConfig(enemyType)
  return { type = enemyType, health = 50, damage = 10, speed = 2.0, score = 100, isBoss = false, isElite = false }
end

function startWave(waveNumber)
  WaveManager.waveNumber = waveNumber
  WaveManager.state = "active"
  WaveManager.enemiesSpawned = 0
  WaveManager.enemiesDefeated = 0
  WaveManager.enemiesRemaining = calculateEnemyCount(waveNumber)
  WaveManager.isBossWave = isBossWave(waveNumber)
  WaveManager.isEliteWave = isEliteWave(waveNumber)
  WaveManager.enemyTypes = generateEnemyTypes(waveNumber)
  return { success = true, waveNumber = waveNumber, enemyCount = WaveManager.enemiesRemaining, isBossWave = WaveManager.isBossWave, isEliteWave = WaveManager.isEliteWave, enemyTypes = WaveManager.enemyTypes }
end

function spawnNextEnemy()
  if WaveManager.enemiesSpawned >= WaveManager.enemiesRemaining then
    return { success = false }
  end
  WaveManager.enemiesSpawned = WaveManager.enemiesSpawned + 1
  return { success = true, enemy = getEnemyConfig("basic"), spawnIndex = WaveManager.enemiesSpawned }
end

function onEnemyDefeated(enemyType)
  WaveManager.enemiesDefeated = WaveManager.enemiesDefeated + 1
  WaveManager.enemiesRemaining = WaveManager.enemiesRemaining - 1
  return { success = true, isWaveComplete = WaveManager.enemiesRemaining <= 0 }
end

function pauseWave()
  WaveManager.state = "paused"
  return true
end

function resumeWave()
  WaveManager.state = "active"
  return true
end

function updateWave(deltaTime)
  if WaveManager.state == "active" then
    WaveManager.elapsedTime = WaveManager.elapsedTime + deltaTime
  end
  return { elapsedTime = WaveManager.elapsedTime, state = WaveManager.state, enemiesRemaining = WaveManager.enemiesRemaining }
end

function resetWaveManager()
  WaveManager = { waveNumber = 1, maxWaves = 10, state = "waiting", enemiesSpawned = 0, enemiesDefeated = 0, enemiesRemaining = 0, elapsedTime = 0, difficulty = "normal" }
  return true
end

function getNextWaveNumber()
  return WaveManager.waveNumber + 1
end

function getWaveScoreMultiplier()
  return 1.0 + WaveManager.waveNumber * 0.1
end
      `;
  }

  getWaveState(): WaveState {
    if (!this.initialized) {
      console.warn('[WaveManager] Not initialized');
      return {
        waveNumber: 1,
        maxWaves: 10,
        currentState: 'waiting',
        enemiesSpawned: 0,
        enemiesDefeated: 0,
        enemiesRemaining: 0,
        elapsedTime: 0,
        difficulty: 'normal',
        isBossWave: false,
        isEliteWave: false,
        progress: 0
      };
    }

    try {
      const stubModule = luaEngine.getStubModule('wave_manager_module');
      if (stubModule) {
        const getWaveStateFunc = (stubModule as Record<string, unknown>)['getWaveState'] as (...args: unknown[]) => unknown;
        if (getWaveStateFunc) {
          const result = getWaveStateFunc();
          if (result && typeof result === 'object') {
            return (result as Record<string, unknown>).state as WaveState;
          }
        }
      }
      return luaEngine.call<WaveState>('getWaveState');
    } catch (error) {
      console.error('[WaveManager] Failed to get wave state:', error);
      return {
        waveNumber: 1,
        maxWaves: 10,
        currentState: 'waiting',
        enemiesSpawned: 0,
        enemiesDefeated: 0,
        enemiesRemaining: 0,
        elapsedTime: 0,
        difficulty: 'normal',
        isBossWave: false,
        isEliteWave: false,
        progress: 0
      };
    }
  }

  setDifficulty(difficulty: DifficultyLevel): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('setDifficulty', difficulty);
    } catch {
      return false;
    }
  }

  setMaxWaves(maxWaves: number): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('setMaxWaves', maxWaves);
    } catch {
      return false;
    }
  }

  calculateEnemyCount(waveNumber: number): number {
    if (!this.initialized) return 5;

    try {
      return luaEngine.call<number>('calculateEnemyCount', waveNumber);
    } catch {
      return 5;
    }
  }

  isBossWave(waveNumber: number): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('isBossWave', waveNumber);
    } catch {
      return false;
    }
  }

  isEliteWave(waveNumber: number): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('isEliteWave', waveNumber);
    } catch {
      return false;
    }
  }

  generateEnemyTypes(waveNumber: number): string[] {
    if (!this.initialized) return [];

    try {
      return luaEngine.call<string[]>('generateEnemyTypes', waveNumber);
    } catch {
      return [];
    }
  }

  getEnemyConfig(enemyType: string): EnemyConfig | null {
    if (!this.initialized) return null;

    try {
      return luaEngine.call<EnemyConfig>('getEnemyConfig', enemyType);
    } catch {
      return null;
    }
  }

  startWave(waveNumber: number): { success: boolean; waveNumber: number; enemyCount: number; isBossWave: boolean; isEliteWave: boolean; enemyTypes: string[] } | { success: boolean; error: string } {
    if (!this.initialized) {
      return { success: false, error: 'not_initialized' };
    }

    try {
      const stubModule = luaEngine.getStubModule('wave_manager_module');
      if (stubModule) {
        const startWaveFunc = (stubModule as Record<string, unknown>)['startWave'] as (...args: unknown[]) => unknown;
        if (startWaveFunc) {
          const result = startWaveFunc(waveNumber);
          return result as { success: boolean; waveNumber: number; enemyCount: number; isBossWave: boolean; isEliteWave: boolean; enemyTypes: string[] } | { success: boolean; error: string };
        }
      }
      return luaEngine.call<{ success: boolean; waveNumber: number; enemyCount: number; isBossWave: boolean; isEliteWave: boolean; enemyTypes: string[] } | { success: boolean; error: string }>('startWave', waveNumber);
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'unknown_error' };
    }
  }

  spawnNextEnemy(): { success: boolean; enemy: EnemyConfig; spawnIndex: number; totalToSpawn: number } | { success: boolean; error: string } {
    if (!this.initialized) {
      return { success: false, error: 'not_initialized' };
    }

    try {
      return luaEngine.call<{ success: boolean; enemy: EnemyConfig; spawnIndex: number; totalToSpawn: number } | { success: boolean; error: string }>('spawnNextEnemy');
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'unknown_error' };
    }
  }

  onEnemyDefeated(enemyType: string): { success: boolean; enemiesDefeated: number; enemiesRemaining: number; isWaveComplete: boolean; score: number } | { success: boolean; error: string } {
    if (!this.initialized) {
      return { success: false, error: 'not_initialized' };
    }

    try {
      const stubModule = luaEngine.getStubModule('wave_manager_module');
      if (stubModule) {
        const onEnemyDefeatedFunc = (stubModule as Record<string, unknown>)['onEnemyDefeated'] as (...args: unknown[]) => unknown;
        if (onEnemyDefeatedFunc) {
          return onEnemyDefeatedFunc(enemyType) as { success: boolean; enemiesDefeated: number; enemiesRemaining: number; isWaveComplete: boolean; score: number } | { success: boolean; error: string };
        }
      }
      return luaEngine.call<{ success: boolean; enemiesDefeated: number; enemiesRemaining: number; isWaveComplete: boolean; score: number } | { success: boolean; error: string }>('onEnemyDefeated', enemyType);
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'unknown_error' };
    }
  }

  pauseWave(): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('pauseWave');
    } catch {
      return false;
    }
  }

  resumeWave(): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('resumeWave');
    } catch {
      return false;
    }
  }

  update(deltaTime: number): { elapsedTime: number; state: string; enemiesRemaining: number } {
    if (!this.initialized) {
      return { elapsedTime: 0, state: 'waiting', enemiesRemaining: 0 };
    }

    try {
      return luaEngine.call<{ elapsedTime: number; state: string; enemiesRemaining: number }>('updateWave', deltaTime);
    } catch {
      return { elapsedTime: 0, state: 'waiting', enemiesRemaining: 0 };
    }
  }

  reset(): boolean {
    if (!this.initialized) return false;

    try {
      return luaEngine.call<boolean>('resetWaveManager');
    } catch {
      return false;
    }
  }

  getNextWaveNumber(): number | null {
    if (!this.initialized) return null;

    try {
      return luaEngine.call<number | null>('getNextWaveNumber');
    } catch {
      return null;
    }
  }

  getWaveScoreMultiplier(): number {
    if (!this.initialized) return 1.0;

    try {
      return luaEngine.call<number>('getWaveScoreMultiplier');
    } catch {
      return 1.0;
    }
  }

  async reloadScript(): Promise<void> {
    console.log('[WaveManager] Reloading wave script...');
    const newScript = await this.loadWaveScript();
    luaEngine.registerModule({ name: 'wave_manager', script: newScript });
    console.log('[WaveManager] Wave script reloaded');
  }

  destroy(): void {
    this.initialized = false;
    console.log('[WaveManager] Destroyed');
  }
}

export const waveManager = new WaveManager();